File: 40.js
function reverse(head) {
    let prev = null, curr = head;
    while (curr) {
        let next = curr.next;
        curr.next = prev;
        prev = curr;
        curr = next;
    }
    return prev;
}

function removeDescendingNodes(head) {
    if (!head || !head.next) return head;

    // Step 1: Reverse
    head = reverse(head);

    let maxVal = head.val;
    let curr = head;

    // Step 2: Remove smaller ones
    while (curr && curr.next) {
        if (curr.next.val < maxVal) {
            curr.next = curr.next.next;
        } else {
            curr = curr.next;
            maxVal = curr.val;
        }
    }

    // Step 3: Reverse back
    return reverse(head);
}

File: 40.py
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

def reverse(head):
    prev = None
    curr = head
    while curr:
        nxt = curr.next
        curr.next = prev
        prev = curr
        curr = nxt
    return prev

def remove_descending_nodes(head):
    if not head or not head.next:
        return head

    # Step 1: Reverse list
    head = reverse(head)

    max_val = head.val
    curr = head

    # Step 2: Remove nodes with value < max_so_far
    while curr and curr.next:
        if curr.next.val < max_val:
            curr.next = curr.next.next  # delete node
        else:
            curr = curr.next
            max_val = curr.val

    # Step 3: Reverse again to restore original order
    return reverse(head)
